File: python/src/threads/store/writer.py
"""The single writer of one branch: `validate_next`, then the fenced conditional append."""

import asyncio
from collections.abc import Callable, Coroutine, Sequence
from dataclasses import dataclass

from threads.log import BranchId, ParseError
from threads.log.digest import sha256_hex
from threads.redaction import SecretInStoredBytesError, published
from threads.reduce import Fold
from threads.reduce.state import HeadRef, ReducedState, reduced_state
from threads.render.artifacts import ReadArtifact
from threads.result import Err, Ok
from threads.store import lease, sql
from threads.store.admit import Admitted, admit, trial
from threads.store.appended import Appended
from threads.store.companion import Companion
from threads.store.conn import CommitUnknownError, Conn
from threads.store.indexing import index_append, known
from threads.store.lines import Draft, Position, stored_secret
from threads.store.verify import StoredEvent
from threads.store.worker import Clock, Worker


@dataclass(frozen=True, slots=True)
class Refusal[E]:
    """A decided append whose decision refused: nothing was appended, and the writer goes on."""

    refusal: E


@dataclass(frozen=True, slots=True)
class DecideTx:
    """What a decided append's decision reads, inside the append's transaction."""

    conn: Conn
    fold: Fold
    """The committed fold the drafts extend: the stored head is its head."""
    now: int
    """The append's clock: every event's time."""
    read: ReadArtifact
    """The store's artifacts, read on its thread: a ref a decision needs (a big reply's text)."""


type Decide[E] = Callable[[DecideTx], Sequence[Draft] | Refusal[E]]
"""Reads the store and builds the drafts, or refuses. It runs on the store's thread and must not
await."""

_REFUSED = ParseError("invalid_request", "the decision refused")
"""Rolls a refused decision back; a Refused never poisons the writer."""

type _Outcome = lease.Batch | ParseError | lease.Refused


@dataclass(frozen=True, slots=True)
class _Commit:
    """One append's commit: where it goes, the trial fold its committed attempt admitted into,
    its clock and companion."""

    head: lease.Head
    fold: Callable[[], Fold]
    now: int
    companion: Companion | None


class Writer:
    """Holds a branch lease at one epoch. Each `append` validates the events against the
    reduced state, then commits them only if the lease is still this writer's and the head is
    where it expects. A lost lease or a moved head poisons the writer."""

    def __init__(  # noqa: PLR0913, PLR0917 - the lease, its fold and head, the clock, the artifacts
        self,
        worker: Worker,
        held: lease.Lease,
        fold: Fold,
        last_line: bytes,
        clock: Clock,
        read: ReadArtifact,
    ) -> None:
        if fold.segment is None or fold.thread_id is None:
            raise ValueError("a writer needs a folded branch")
        self._worker = worker
        self._lease = held
        self._fold = fold
        self._branch: BranchId = fold.segment
        self._last_line = last_line
        self._clock = clock
        self._read = read
        self._poisoned = False
        self._lock = asyncio.Lock()
        self._moved = asyncio.Event()
        effects = fold.effects.values()
        in_doubt = any(s == "unknown" for _, s in effects)
        self._requires_recovery = bool(fold.pending or fold.open_requests) or in_doubt

    @property
    def branch_id(self) -> BranchId:
        return self._branch

    @property
    def epoch(self) -> int:
        return self._lease.epoch

    @property
    def owner(self) -> lease.Owner:
        """What fences this writer's resource-ledger rows."""
        return lease.Owner(self._branch, self._lease)

    @property
    def fold(self) -> Fold:
        """The committed branch folded through its last append. Read it; never mutate it."""
        return self._fold

    def state(self) -> ReducedState:
        """ReducedState at the committed head: what hooks are shown."""
        return reduced_state(self._fold, HeadRef(self._fold.seq, sha256_hex(self._last_line)))

    @property
    def requires_recovery(self) -> bool:
        """True when the branch had a call without a result, a model request without a response,
        or an effect in doubt when this writer took it. Normal dispatch must refuse such a
        writer: only recovery may dispatch on it, after re-checking approval,
        cancellation and policy."""
        return self._requires_recovery

    async def append(
        self,
        drafts: Sequence[Draft],
        companion: Companion | None = None,
        *,
        admit: Callable[[Fold, Sequence[Draft]], Sequence[Draft]] | None = None,
    ) -> Ok[tuple[StoredEvent, ...]] | Err[ParseError]:
        """Appends the drafts as one transaction and resolves after it is durable. A companion
        that refuses rolls the append back: its error is the result and the writer goes on.
        `admit` picks the drafts the batch keeps, from the fold as it is under the writer's lock
        (another task's append can land while this one waits for it). The result is what was
        appended; `Runtime.append_with` reports a batch `admit` changed as `Barred`."""
        async with self._lock:
            if self._poisoned:
                return Err(ParseError("writer_poisoned", "this writer lost its lease or head"))
            chosen = tuple(drafts if admit is None else admit(self._fold, drafts))
            if not chosen:
                return Ok(())

            def fixed(_tx: DecideTx) -> Sequence[Draft] | Refusal[None]:
                return chosen

            outcome, _ = await self._run(fixed, companion)
            return _result(outcome)

    async def append_decided[E](
        self, decide: Decide[E]
    ) -> Ok[tuple[StoredEvent, ...]] | Err[ParseError] | Refusal[E]:
        """One transaction that checks the lease and head, lets `decide` read the store and
        build the drafts, admits them (on the store's thread) and commits. A refusal rolls back
        and leaves the writer usable; `stale_epoch` and `seq_conflict` poison it, as for
        `append`. A batch that comes out empty commits nothing and moves nothing."""
        async with self._lock:
            if self._poisoned:
                return Err(ParseError("writer_poisoned", "this writer lost its lease or head"))
            outcome, refusal = await self._run(decide, None)
            return refusal if refusal is not None else _result(outcome)

    async def _run[E](
        self, decide: Decide[E], companion: Companion | None
    ) -> tuple[_Outcome, Refusal[E] | None]:
        """The append's transaction: the fence, then the decision and the admission on the
        store's thread, into a trial copy of the fold that replaces it once committed."""
        now = self._clock()
        head = lease.Head(self._branch, self._lease, self._fold.seq)
        committed = self._fold
        # One entry per attempt that reached the decision: the store may re-run the whole
        # transaction, so each attempt admits into its own trial of the committed fold, and
        # only the last attempt (the one that committed) counts.
        tries: list[tuple[Fold, Refusal[E] | None]] = []

        def build(conn: Conn) -> lease.Batch | lease.Refused:
            fold = trial(committed)
            decided = decide(DecideTx(conn, committed, now, self._read))
            tries.append((fold, decided if isinstance(decided, Refusal) else None))
            if isinstance(decided, Refusal):
                return lease.Refused(_REFUSED)
            built = self._admit(fold, decided, now)
            if isinstance(built, Err):
                return lease.Refused(built.error)
            return self._batch(head, built.value)

        def last() -> Fold:
            return tries[-1][0] if tries else committed

        outcome = await self._commit(_Commit(head, last, now, companion), build)
        refused = isinstance(outcome, lease.Refused) and len(tries) > 0
        return outcome, tries[-1][1] if refused else None

    def _admit(
        self, fold: Fold, drafts: Sequence[Draft], now: int
    ) -> Ok[Admitted] | Err[ParseError]:
        thread = fold.thread_id
        if thread is None:
            raise ValueError("a writer needs a folded branch")
        at = Position(thread, self._branch, fold.seq + 1, self.epoch, self._last_line, now)
        return admit(fold, drafts, at)

    def _batch(self, head: lease.Head, admitted: Admitted) -> lease.Batch:
        rows = admitted.rows
        last = rows[-1][1] if rows else self._last_line
        return lease.Batch(head.expected_seq, rows, sha256_hex(last), admitted.opened)

    async def _commit(self, commit: _Commit, build: lease.Build) -> _Outcome:
        """Runs the append, its decision included, to settlement even if the caller is
        cancelled: the statement can't be recalled once queued. Until it settles the writer is
        poisoned, so a second cancellation leaves it poisoned, never out of step."""
        self._poisoned = True
        head, now, after = commit.head, commit.now, self._after(commit)
        op = asyncio.ensure_future(
            self._worker.call(
                lambda c: _published(lambda: lease.append(c, head, now, build, after))
            )
        )
        try:
            outcome = await asyncio.shield(op)
        except asyncio.CancelledError:
            self._settle(await op, commit.fold())
            raise
        self._settle(outcome, commit.fold())
        return outcome

    def _settle(self, outcome: _Outcome, fold: Fold) -> None:
        """A committed batch replaces the fold and moves the writer; an empty one or a refusal
        leaves both; a lost lease or head leaves it poisoned."""
        if isinstance(outcome, ParseError):
            return
        self._poisoned = False
        if isinstance(outcome, lease.Batch) and outcome.rows:
            self._fold = fold
            self._last_line = outcome.rows[-1][1]
            self._moved.set()
            self._moved = asyncio.Event()

    def _after(self, commit: _Commit) -> lease.After:
        """The index hooks over the append's events, then its companion."""
        holder, companion = self._lease.holder_id, commit.companion

        def after(conn: Conn, stored: sql.Branch, batch: lease.Batch) -> ParseError | None:
            events = [event for event, _ in batch.rows]
            appended = Appended(
                stored.tenant_id,
                stored.thread_id,
                self._branch,
                known(events),
                batch.opened,
                holder,
                commit.now,
            )
            indexed = index_append(conn, appended)
            if indexed is not None or companion is None:
                return indexed
            return companion(conn, events)

        return after

    def moved(self) -> Coroutine[None, None, bool]:
        """Returns on this writer's next committed append after this call, whoever made it (a
        control included)."""
        return self._moved.wait()

    async def fence(self) -> Ok[None] | Err[ParseError]:
        """Checked immediately before anything is dispatched (a model attempt, a tool body, a
        lookup, a termination): a writer whose lease moved on must not reach the adapter, even
        though its intent is already durable. A lost lease poisons the writer."""
        now = self._clock()
        error = await self._worker.read(lambda c: lease.check(c, self._branch, self._lease, now))
        if error is not None:
            self._poisoned = True
            return Err(error)
        return Ok(None)

    async def release(self) -> None:
        """Hands the lease back so the next executor can take the branch at once, only while
        this holder and epoch still hold it. The writer is done afterwards."""
        now = self._clock()
        self._poisoned = True
        await self._worker.call(lambda c: lease.release(c, self._branch, self._lease, now))

    async def renew(self) -> Ok[None] | Err[ParseError]:
        """Extends the lease. Once lost it stays lost: the writer is poisoned."""
        now = self._clock()
        try:
            renewed = await self._worker.call(
                lambda c: lease.renew(c, self._branch, self._lease, now)
            )
        except CommitUnknownError:
            # The renewal may or may not have landed: the owner reloads from the log.
            self._poisoned = True
            raise
        if isinstance(renewed, ParseError):
            self._poisoned = True
            return Err(renewed)
        self._lease = renewed
        return Ok(None)


def _result(outcome: _Outcome) -> Ok[tuple[StoredEvent, ...]] | Err[ParseError]:
    if isinstance(outcome, lease.Refused):
        return Err(outcome.error)
    if isinstance(outcome, ParseError):
        return Err(outcome)
    return Ok(tuple(event for event, _ in outcome.rows))


def _published(append: Callable[[], _Outcome]) -> _Outcome:
    """The append with secret registration paused until its rows are durable (C5): each draft's
    content is checked as it is admitted, inside, so no value registered after that check reaches
    the store."""
    try:
        return published(b"", append)
    except SecretInStoredBytesError:
        return lease.Refused(stored_secret())

File: typescript/packages/core/src/store/writer.ts
import { err, ok, type Result } from "../result";
import type { Chain, ChainEvent, VerifiedLog } from "../verify";
import { type LogError, logError } from "../verify/error";
import { VERSION } from "../version";
import {
  type Admitted,
  admitDrafts,
  copyChain,
  type EventDraft,
} from "./admit";
import { recordApprovals } from "./approvals";
import { CommitUnknown, READ_ONLY, type StoreDriver, type Tx } from "./driver";
import { indexAppend, knownOf } from "./indexing";
import {
  atomically,
  type BranchRow,
  getBranch,
  getLease,
  insertEvents,
  putLease,
} from "./tables";
import { Fifo } from "./tx";

export type { EventDraft } from "./admit";

export type Lease = {
  readonly branchId: string;
  readonly holderId: string;
  readonly epoch: number;
  /** How long the lease lasts from each take or renewal. */
  readonly ttlMs: number;
};

/**
 * What a decided append's `decide` reads, inside the append's transaction. `decide` may run
 * again from the start (see `Tx`): it reads only through `tx` and returns what it built.
 */
export type DecideTx = {
  readonly tx: Tx;
  /** The committed chain the drafts extend: the stored head is its head. */
  readonly chain: Chain;
  /** The append's clock: every event's time. */
  readonly now: number;
};

/**
 * Host rows written in an append's transaction, after its events and index rows (a receipt, a
 * consumed inbox item). It may run again from the start, and never calls back into the writer.
 */
export type Alongside = (
  added: readonly ChainEvent[],
  tx: Tx,
) => Promise<Result<void, LogError>>;

/** A decided append whose decision refused: nothing was appended, and the writer goes on. */
export type Refusal<E> = { readonly kind: "refused"; readonly refusal: E };

/** Whether a decided append was refused, rather than appended or failed. */
export function isRefusal<T, E>(
  outcome: T | Refusal<E>,
): outcome is Refusal<E> {
  return (
    typeof outcome === "object" &&
    outcome !== null &&
    "kind" in outcome &&
    outcome.kind === "refused"
  );
}

/** The `writer.impl` this implementation writes in every header it creates. */
export const IMPL = "threads-ts";

const major = (version: string): string => version.split(".")[0] ?? "";

/**
 * Only the implementation a branch's header names, at the same major version, appends to it
 *. Anyone may read, export or fork it.
 */
export function writerMismatch(log: VerifiedLog): LogError | undefined {
  const writer = log.segments.at(-1)?.header.writer;
  if (writer?.impl === IMPL && major(writer.version) === major(VERSION))
    return undefined;
  const named = `${writer?.impl} ${writer?.version}`;
  return logError(
    "writer_mismatch",
    `${named} writes this branch; continue on a fork`,
    0,
  );
}

// Rolls a refused decision back; its code never poisons the writer.
const REFUSED: LogError = logError("invalid_request", "the decision refused");

const poisoned = (): LogError =>
  logError("writer_poisoned", "this writer lost its lease or head");

type Produce = (
  tx: DecideTx,
) => Promise<Result<readonly EventDraft[], LogError>>;

/**
 * The single writer of one branch under one lease epoch. Every append runs the
 * same admission as import, then commits only if the lease is still this one and the stored
 * head is where this writer left it. A stale lease or a moved head poisons the writer, and so
 * does a commit whose outcome is unknown: the owner reloads the branch from the log.
 */
export class Writer {
  readonly #db: StoreDriver;
  readonly #now: () => number;
  /** One append, fence, renewal or release at a time, each against the committed chain. */
  readonly #lock = new Fifo();
  readonly lease: Lease;
  /**
   * The chain had a pending tool call, a model request awaiting its response, or an effect
   * `begun` or `unknown` when this lease was taken. Such a branch is in doubt: normal dispatch must refuse this writer, and only
   * recovery may settle and continue it.
   */
  readonly requiresRecovery: boolean;
  #chain: Chain;
  #poisoned = false;
  #moved = Promise.withResolvers<void>();

  constructor(db: StoreDriver, now: () => number, lease: Lease, chain: Chain) {
    this.#db = db;
    this.#now = now;
    this.lease = lease;
    this.#chain = chain;
    const { pending, awaiting, effects } = chain.fold;
    this.requiresRecovery =
      pending.size > 0 ||
      awaiting.size > 0 ||
      effects
        .values()
        .some((e) => e.status === "begun" || e.status === "unknown");
  }

  /** Settles on this writer's next committed append, whoever made it (a control included). */
  moved(): Promise<void> {
    return this.#moved.promise;
  }

  /** The committed chain this writer appends to. Each append replaces it; none mutates it. */
  get chain(): Chain {
    return this.#chain;
  }

  /**
   * Appends drafts in order and returns once their transaction has committed with full sync.
   * `alongside` writes host rows in the same transaction; an error from it rolls the append back.
   */
  append(
    drafts: readonly EventDraft[],
    alongside?: Alongside,
  ): Promise<Result<readonly ChainEvent[], LogError>> {
    return this.#locked(() =>
      this.#run(this.#db, async () => ok(drafts), alongside),
    );
  }

  /**
   * `append` as a savepoint of `outer`, a transaction the caller holds; the chain moves once
   * `outer` commits. For a writer nobody else appends through yet (a fork's, a repair's).
   */
  appendIn(
    outer: Tx,
    drafts: readonly EventDraft[],
  ): Promise<Result<readonly ChainEvent[], LogError>> {
    return this.#locked(() => this.#run(outer, async () => ok(drafts)));
  }

  /**
   * One transaction that begins, checks the lease and head, lets `decide` read the store and
   * build the drafts, admits them and commits. A refusal rolls back and leaves the writer
   * usable; `stale_epoch` and `seq_conflict` poison it, as for `append`. A batch that comes out
   * empty (the barrier dropped every draft) commits nothing and moves nothing.
   */
  async appendDecided<E>(
    decide: (tx: DecideTx) => Promise<Result<readonly EventDraft[], E>>,
    alongside?: Alongside,
  ): Promise<Result<readonly ChainEvent[], LogError> | Refusal<E>> {
    // Each attempt decides afresh: only the last attempt's refusal counts.
    let refusal: Refusal<E> | undefined;
    const appended = await this.#locked(() =>
      this.#run(
        this.#db,
        async (tx) => {
          refusal = undefined;
          const decided = await decide(tx);
          if (decided.ok) return decided;
          refusal = { kind: "refused", refusal: decided.error };
          return err(REFUSED);
        },
        alongside,
      ),
    );
    return refusal ?? appended;
  }

  async #locked<T>(fn: () => Promise<T>): Promise<T> {
    const release = await this.#lock.turn();
    try {
      return await fn();
    } finally {
      release();
    }
  }

  async #run(
    sql: StoreDriver | Tx,
    produce: Produce,
    alongside?: Alongside,
  ): Promise<Result<readonly ChainEvent[], LogError>> {
    if (this.#poisoned) return err(poisoned());
    const committed = await this.#guard(() =>
      atomically(sql, (tx) => this.#attempt(tx, produce, alongside)),
    );
    if (!committed.ok) {
      const { code } = committed.error;
      this.#poisoned = code === "stale_epoch" || code === "seq_conflict";
    }
    return committed;
  }

  /** One attempt, built from the committed chain, so a retried attempt starts clean. */
  async #attempt(
    tx: Tx,
    produce: Produce,
    alongside: Alongside | undefined,
  ): Promise<Result<readonly ChainEvent[], LogError>> {
    const trial = copyChain(this.#chain);
    const now = this.#now();
    const branch = await this.#fencedHead(tx);
    if (!branch.ok) return branch;
    const drafts = await produce({ tx, chain: this.#chain, now });
    if (!drafts.ok) return drafts;
    const admitted = admitDrafts(trial, drafts.value, now, this.lease.epoch);
    if (!admitted.ok) return admitted;
    const { events } = admitted.value;
    const written = await this.#write(tx, branch.value, admitted.value, now);
    if (!written.ok) return written;
    const host =
      alongside === undefined ? ok(undefined) : await alongside(events, tx);
    if (!host.ok) return host;
    if (events.length > 0) tx.afterCommit(() => this.#publish(trial));
    return ok(events);
  }

  /** A committed append moves the chain and wakes whoever waits on this writer. */
  #publish(trial: Chain): void {
    this.#chain = trial;
    this.#moved.resolve();
    this.#moved = Promise.withResolvers<void>();
  }

  /** A commit whose outcome is unknown poisons the writer: its owner reloads it from the log. */
  async #guard<T>(fn: () => Promise<T>): Promise<T> {
    try {
      return await fn();
    } catch (error) {
      if (error instanceof CommitUnknown) this.#poisoned = true;
      throw error;
    }
  }

  /**
   * The dispatch fence: this writer still holds the lease at its epoch, read
   * from the store right before an adapter is called. A stale writer is poisoned and must never
   * dispatch. Passing proves nothing about what an older owner already sent.
   */
  fence(): Promise<Result<void, LogError>> {
    return this.#locked(async () => {
      if (this.#poisoned) return err(poisoned());
      const live = await this.#db.transaction(
        (tx) => this.#checkLease(tx),
        READ_ONLY,
      );
      if (!live.ok) this.#poisoned = true;
      return live;
    });
  }

  /**
   * Runs `fn` in one transaction that first checks this lease, so a store write outside the log
   * (a resource ledger row) is fenced like an append. A stale writer is poisoned.
   */
  async fenced<T>(
    fn: (tx: Tx) => Promise<Result<T, LogError>>,
  ): Promise<Result<T, LogError>> {
    if (this.#poisoned) return err(poisoned());
    const done = await atomically(this.#db, async (tx) => {
      const live = await this.#checkLease(tx);
      return live.ok ? fn(tx) : live;
    });
    if (!done.ok && done.error.code === "stale_epoch") this.#poisoned = true;
    return done;
  }

  /** Extends the lease; fails (and poisons) if another holder or epoch took it or it lapsed. */
  renew(ttlMs: number): Promise<Result<void, LogError>> {
    return this.#locked(async () => {
      if (this.#poisoned) return err(poisoned());
      const renewed = await this.#guard(() =>
        atomically(this.#db, async (tx) => {
          const live = await this.#checkLease(tx);
          if (!live.ok) return live;
          await this.#putLease(tx, this.#now() + ttlMs);
          return ok(undefined);
        }),
      );
      if (!renewed.ok) this.#poisoned = true;
      return renewed;
    });
  }

  /**
   * Hands the lease back so the next executor can take the branch at once. Only while this
   * holder and epoch still hold it: a stale writer never clears a newer owner's lease. Only the
   * lease changes; in-doubt work stays in the log for recovery. The writer is done afterwards.
   */
  release(): Promise<void> {
    return this.#locked(async () => {
      const live = !this.#poisoned;
      this.#poisoned = true;
      if (!live) return;
      await atomically(this.#db, async (tx) => {
        if ((await this.#checkLease(tx)).ok)
          await this.#putLease(tx, this.#now());
        return ok(undefined);
      });
    });
  }

  #putLease(tx: Tx, expiresAt: number): Promise<void> {
    return putLease(tx, this.lease.branchId, {
      holder_id: this.lease.holderId,
      epoch: this.lease.epoch,
      expires_at: expiresAt,
    });
  }

  /** The lease is still this one and the stored head is where this writer left it. */
  async #fencedHead(tx: Tx): Promise<Result<BranchRow, LogError>> {
    const live = await this.#checkLease(tx);
    if (!live.ok) return live;
    const branch = await getBranch(tx, this.lease.branchId);
    if (!branch.ok) return branch;
    const expected = this.#chain.fold.seq;
    if (branch.value === undefined || branch.value.head_seq !== expected)
      return err(
        logError("seq_conflict", `the stored head is not ${expected}`),
      );
    return ok(branch.value);
  }

  /** The admitted events, their index rows and their approvals' challenge rows. */
  async #write(
    tx: Tx,
    branch: BranchRow,
    admitted: Admitted,
    now: number,
  ): Promise<Result<void, LogError>> {
    const { events, opened } = admitted;
    await insertEvents(tx, this.lease.branchId, events);
    const own = this.#chain.segments.at(-1)?.header;
    if (own === undefined) throw new Error("a writer's chain has a header");
    const indexed = await indexAppend({
      tx,
      tenant: branch.tenant_id,
      threadId: own.thread_id,
      branchId: own.branch_id,
      events: knownOf(events),
      opened,
      holderId: this.lease.holderId,
      now,
    });
    if (!indexed.ok) return indexed;
    return recordApprovals(tx, branch, events, now);
  }

  async #checkLease(tx: Tx): Promise<Result<void, LogError>> {
    const lease = await getLease(tx, this.lease.branchId);
    if (!lease.ok) return lease;
    const row = lease.value;
    const current =
      row !== undefined &&
      row.holder_id === this.lease.holderId &&
      row.epoch === this.lease.epoch &&
      row.expires_at > this.#now();
    return current
      ? ok(undefined)
      : err(
          logError(
            "stale_epoch",
            `epoch ${this.lease.epoch} no longer holds the lease`,
          ),
        );
  }
}
